src/java.base/share/classes/sun/security/util/PBEUtil.java

```diff
@@ -0,0 +1,355 @@
+                            "and iteration count");
+                }
+            } else if (!(params instanceof PBEParameterSpec)) {
+                throw new InvalidAlgorithmParameterException(
+                        "PBEParameterSpec type required");
+            } else {
+                PBEParameterSpec pbeParams = (PBEParameterSpec) params;
+                // make sure the parameter values are consistent
+                if (salt != null) {
+                    if (!Arrays.equals(salt, pbeParams.getSalt())) {
+                        throw new InvalidAlgorithmParameterException(
+                                "Inconsistent value of salt " +
+                                "between key and params");
+                    }
+                } else {
+                    salt = pbeParams.getSalt();
+                }
+                if (iCount != 0) {
+                    if (iCount != pbeParams.getIterationCount()) {
+                        throw new InvalidAlgorithmParameterException(
+                                "Different iteration count " +
+                                "between key and params");
+                    }
+                } else {
+                    iCount = pbeParams.getIterationCount();
+                }
+            }
+            // For security purpose, we need to enforce a minimum length
+            // for salt; just require the minimum salt length to be 8-byte
+            // which is what PKCS#5 recommends and openssl does.
+            if (salt.length < 8) {
+                throw new InvalidAlgorithmParameterException(
+                        "Salt must be at least 8 bytes long");
+            }
+            if (iCount <= 0) {
+                throw new InvalidAlgorithmParameterException(
+                        "IterationCount must be a positive number");
+            }
+            return new PBEKeySpec(passwdChars, salt, iCount);
+        } finally {
+            Arrays.fill(passwdChars, '\0');
+        }
+    }
+
+    /*
+     * Check that the key implements the PBEKey interface. If params is an
+     * instance of PBEParameterSpec, validate consistency with the key's
+     * derivation data. Used by P11Mac and P11PBECipher (SunPKCS11).
+     */
+    public static void checkKeyAndParams(Key key,
+            AlgorithmParameterSpec params, String algorithm)
+            throws InvalidKeyException, InvalidAlgorithmParameterException {
+        if (key instanceof javax.crypto.interfaces.PBEKey pbeKey) {
+            if (params instanceof PBEParameterSpec pbeParams) {
+                if (pbeParams.getIterationCount() !=
+                        pbeKey.getIterationCount() ||
+                        !Arrays.equals(pbeParams.getSalt(), pbeKey.getSalt())) {
+                    throw new InvalidAlgorithmParameterException(
+                            "Salt or iteration count parameters are " +
+                            "not consistent with PBE key");
+                }
```

Comment:
Based on the javadoc of PBEKey, it's possible that salt and ic aren't specified, e.g. null and 0. In this case, this check will fail if the PBEParameterSpec contains salt and ic. IIRC, the java impl will use the values from PBEParameterSpec in this case. Perhaps we can consider same handling? This can be handled separately by tracking this under a separate bug as it's getting too close to RPD1 already.

Reply:
The PBEUtil::checkKeyAndParams function is used by SunPKCS11 only (see P11PBECipher::engineInit and P11Mac::engineInit), and the keys passed are always P11Key instances. Thus, a P11Key that is also a PBEKey is a P11PBEKey, and both the salt and iteration count for these keys are properly set.